mlflow/server/js/src/components/ExperimentPage.js

```diff
@@ -16,49 +18,83 @@ class ExperimentPage extends Component {
     super(props);
     this.onSearch = this.onSearch.bind(this);
     this.getRequestIds = this.getRequestIds.bind(this);
+    const store = ExperimentPage.getLocalStore(this.props.experimentId);
+    // Load state data persisted in localStorage. If data isn't present in localStorage (e.g. the
+    // first time we construct this component in a browser), the default values in
+    // ExperimentPagePersistedState will take precedence.
+    const persistedState = new ExperimentPagePersistedState(store.loadComponentState());
+    this.state = {
+      ...ExperimentPage.getDefaultUnpersistedState(),
+      ...persistedState.toJSON(),
+    };
   }
 
   static propTypes = {
     experimentId: PropTypes.number.isRequired,
     dispatchSearchRuns: PropTypes.func.isRequired,
   };
 
-  state = {
-    paramKeyFilter: new KeyFilter(),
-    metricKeyFilter: new KeyFilter(),
-    getExperimentRequestId: getUUID(),
-    searchRunsRequestId: getUUID(),
-    searchInput: '',
-    lastExperimentId: undefined,
-    lifecycleFilter: LIFECYCLE_FILTER.ACTIVE,
-  };
+  /** Returns default values for state attributes that aren't persisted in local storage. */
+  static getDefaultUnpersistedState() {
+    return {
+      // String UUID associated with a GetExperiment API request
+      getExperimentRequestId: getUUID(),
+      // String UUID associated with a SearchRuns API request
+      searchRunsRequestId: getUUID(),
+      // Last experiment, if any, displayed by this instance of ExperimentPage
+      lastExperimentId: undefined,
+      // Lifecycle filter of runs to display
+      lifecycleFilter: LIFECYCLE_FILTER.ACTIVE,
+    };
+  }
+
+  /**
+   * Returns a LocalStorageStore instance that can be used to persist data associated with the
+   * ExperimentPage component (e.g. component state like metric/param filter info), for the
+   * specified experiment.
+   */
+  static getLocalStore(experimentId) {
+    return LocalStorageUtils.getStore("ExperimentPage", experimentId);
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.experimentId === prevProps.experimentId) {
```

Comment:
This is subtle -- I thought the store is keyed based on experimentId?

Reply:
Yep, the store is indeed keyed on experimentId, hence why we create a new `LocalStorageStore` instance [here](https://github.com/mlflow/mlflow/pull/687/files/7e5b76b31ceb31efc7a3f41113576a7729c531ce#diff-8a79a4fb60156a01191f5ed474381086R63)

For context I ran into this when toggling between experiments - it turns out that when `props.experimentId` changes React will reuse the ExperimentPage instance as described [here](https://reactjs.org/docs/reconciliation.html#component-elements-of-the-same-type), and invoke the sequence of update methods described [here](https://reactjs.org/docs/react-component.html#updating) on the component.  The main important thing here is that we don't initialize the store once in the constructor & instead recreate it with the most up-to-date experiment ID (otherwise we'd risk persisting the new experiment's state, contained in `this.state`, to a previous experiment). The `if` check here is just a small optimization to prevent us from snapshotting the state when switching experiments (since switching to a new experiment doesn't change the state of the old experiment), but isn't strictly necessary.

Reply:
I see. If it's just an optimization, maybe we can document that (and/or remove it). I thought it was a correctness issue.